### PR TITLE
Add draft ABI spec

### PR DESCRIPTION
**Pull request by Taction**
This can be a draft idea that moves forwards.

### CODE REVIEW

**Comment by codefromthecrypt** [comment]
one thing I feel we should do is keep this open for holistic view, but spin out functions after trying them in tinygo/fasthttp

I made a skeleton project for tinygo only having log functionality and soon will make fasthttp so we can integrate and test things https://github.com/http-wasm/http-wasm-guest-tinygo/pull/1

**Comment by codefromthecrypt** [comment]
I ran out of time and have some personal time off until monday. I'll definitely put the fasthttp up monday including some realistic benchmark usecase. It won't do anything like pooling for start, mainly make abi sketching more concrete.

**Comment by pkedy** [comment]
@Taction @codefromthecrypt what are your thoughts on the performance characteristics between having several fine-grained calls in the ABI or passing at least all the headers so there are fewer hops through the FFI boundary? I created a example of the latter here: https://github.com/wapc/examples/tree/main/middleware

File: SPEC.md [hunk 1]
```diff
@@ -0,0 +1,172 @@
...
+  - none
+- returns:
+  - none
+
+Start function which is called when the module is loaded and initialized. This can be used by SDKs to setup and/or initialize state, but no proxy_ functions can be used at that point yet.
+
+### `proxy_on_request`
+
+- params:
+  - `i32 (uint32_t) context_id`
```
Comment (codefromthecrypt) [inline SPEC.md:19]: I think this should be a uint64 as many context IDs are that length (including fasthttp)

**Comment by codefromthecrypt** [comment]
all: update there are a lot of variables in how middleware work and also dapr thread leading to this noted interest in net/http. I think it is better to make a -host-go monorepo which can help direct which logic goes where and also ease iteration. Abstractions without at least three implementors are usually busted, so having at least net/http and fasthttp there will help in straw man proposals.

@pkedy on perf impact of fine grained calls. benchmarks benchmarks benchmarks! at least in go, we can do this and I know this is how @inkeliz [Karmem](https://github.com/inkeliz/karmem/tree/master/benchmark) decides on how and if to change things. Sometimes things are surprising, and also sometimes things move. Ex a change in the runtime makes something otherwise not performant better, obviating strange workarounds.

To do this, we need a few cases which are small but representative. Ex.
* tracing headers (implies set_header multiple times usually)
* auth request (regex applied to path or whatever)

once a first impl is in, someone can make a branch with an alternate design and have a before/after benchmark. I think of this both for granularity and also for where the lifecycle works. Ex which code is defined in the guest vs host functions and which scope things operate at.


Concretely, in wazero until we added our [vs](https://github.com/tetratelabs/wazero/tree/main/internal/integration_test/vs) benchmarks we had no way to verify some design change was relevant in terms of perf or safety. I think the same things apply here and main thing is getting in the infrastructure to objectively evaluate designs. This concedes that the first design may be suboptimal and that's fine, as long as we can be objective about changing it!

**Comment by codefromthecrypt** [comment]
update: realizing this is the end of the month and wazero promises monthly releases, I got a little displaced progressing things. I have a draft based on my earlier comment, but it is incomplete. That said, I should have something raised by this weekend. There are a lot of thoughts that are better placed in the code, so I'll leave them for there vs describe now.

**Comment by codefromthecrypt** [comment]
here's a draft of the general design for synchronous handlers (interceptor pattern). this intentionally doesn't get into very specific details like logging levels or which http properties to expose as that allows a foundation to be written.

https://github.com/http-wasm/http-wasm-host-go/pull/1

after this is in, I'd add fasthttp, then benchmarks. finally with this infrastructure in place, we can bat around points like how to represent each HTTP property as wasm and know where to doc, test and bench each part. That's when the infrastructure work will pay off as it will help ensure adding more things != debt or lack of insight into perf characteristics.

my 2p!

**Comment by codefromthecrypt** [comment]
this adds MVP ABI exports needed to put together a demo comparing approaches. after merged, I still need to add fasthttp support and actually plumb a dapr example, as well update tinygo and this repo with the ABI changes. Once all of that's together, all the infrastructure is ready for folks to take this towards whatever end.

https://github.com/http-wasm/http-wasm-host-go/pull/2

**Comment by codefromthecrypt** [comment]
update: have a PR to add fasthttp which proves at least two impls can work. After this is merged, a little polish then I'll setup the comparison tests. https://github.com/http-wasm/http-wasm-host-go/pull/3

I'm traveling (going to golab.io), so will be sporadically available, but we're close and can kick dapr around within the next few days.

**Comment by codefromthecrypt** [comment]
update: I've started dapr integration doing the same as the existing middleware, except using the work-in-progress host/guest code. I'll update again once various PRs are merged, and there's a more sophisticated example as well benchmarks

https://github.com/dapr/components-contrib/compare/master...http-wasm:components-contrib:httpwasm

**Comment by codefromthecrypt** [comment]
alright I'm going to hand this over to @anuraaga now, though he's not completely online. I think the results are promising, that the current impl can get good performance while also allowing flexibility in what users can do. By flexibility I mean they can code the entire handler in a natural programming language, not a DSL or a specific special-cased function. I'll be watching for anything and happy to facilitate further.

**Comment by codefromthecrypt** [comment]
The current repository docs are synced with the [components-contrib fork](https://github.com/http-wasm/components-contrib/blob/httpwasm/middleware/http/wasm/benchmark_test.go) demonstration.

However, there is some more work needed before we can integrate. Notably, we need to be able to rewrite the response body. The below PR is needed to start pulling this apart:

https://github.com/http-wasm/http-wasm-host-go/pull/11

Once this is in I'll add a way to instruct the host to buffer any response body created by the next handler, to prevent it from being eagerly released. This would be an opt-in feature controlled by the guest, and should only be used in the rare case you need to rewrite the response (ex. to drop it or scrub secrets from it).

I expect this done early monday or tuesday, which should be an interesting time, especially as dapr moves to net/http instead of fasthttp for middleware.

**Comment by codefromthecrypt** [comment]
updated the http-wasm fork for components-contrib to not use fasthttp due to upstream changes by @ItalyPaleAle. comparison benchmarks still do until that's swapped out.

**Comment by codefromthecrypt** [comment]
actually I'll swap out fasthttp from benchmarks as it is distracting

**Comment by codefromthecrypt** [comment]
added some important notes on memory allocation as decoupling (yet still allowing) WASI means we have to define how the guest can read fields from the host. https://github.com/http-wasm/http-wasm-abi/pull/13 has a working approach.

Note: decoupling from WASI is important for reasons including WASI itself is a spec in progress, with incompatible functions on the way. We can't rely on WASI for passing configuration to guests (via env or args)

Ex TinyGo will be compilable to wasm without WASI soon, and not all functions require WASI to operate. We'd expect hosts to support current WASI, but our ABI should work regardless. https://github.com/tinygo-org/tinygo/pull/3072

**Comment by codefromthecrypt** [comment]
https://github.com/http-wasm/http-wasm-abi/blob/main/http-handler/http-handler.md

fyi markdown renders in GitHub but not as polished as Hugo. I'll be adding a site shortly, as particularly I'd like to diagram some things.

**Comment by codefromthecrypt** [comment]
fyi here's the first integration for dapr

https://github.com/dapr/components-contrib/pull/2239

I'm happy to help with mosn, just there is some remedial work to do that may take some time first. I think mosn should have latest fasthttp with trailer support before adding wasm to it https://github.com/mosn/pkg/pull/67 I'll continue to help raise PRs and such, but there is a lot of work to do so help wanted!

**Pull request closed without merging**